tests/ui/manual_map_option_2.rs

```diff
@@ -69,3 +69,27 @@ fn main() {
         None => None,
     };
 }
+
+mod issue_12659 {
+    trait DummyTrait {}
+
+    fn foo<T: DummyTrait, F: Fn() -> Result<T, ()>>(f: F) {
+        // Don't lint
+        let _: Option<Result<Box<dyn DummyTrait>, ()>> = match Some(0) {
+            Some(_) => Some(match f() {
+                Ok(res) => Ok(Box::new(res)),
+                _ => Err(()),
+            }),
+            None => None,
+        };
```

Comment:
The problem is that it doesn't catch the adjustment rather than the presence of a `dyn`, e.g. the same can occur with an unsized coercion

```rust
fn main() {
    let x: &[u8; 4] = b"1234";

    // ok
    let _: Option<Box<&[u8]>> = match Some(()) {
        Some(_) => Some(Box::new(x)),
        None => None,
    };

    // error[E0308]: mismatched types
    let _: Option<Box<&[u8]>> = Some(()).map(|_| Box::new(x));
}
```

Reply:
Oops, you're right, I'll see what I can do with it

Reply:
@Alexendoo forgive me for the ping, I'm afraid I got stuck on this. seems like that sometimes its fine with type adjustments, and sometimes not, can this happen other than inside of `Box`?